rfc/text/advanced/ap_authentication_scram.md

```diff
@@ -0,0 +1,387 @@
+              "i=" + Decimal(CHALLENGE.Details.iterations)
+
+CBindName = AUTHENTICATE.Extra.channel_binding
+
+CBindData = AUTHENTICATE.Extra.cbind_data
+
+CBindFlag = IfNull(CBindName, "n", "p=" + CBindName)
+
+CBindInput = CBindFlag + ",," +
+             IfNull(CBindData, "", UnBase64(CBindData))
+
+ClientFinalNoProof = "c=" + Base64(CBindInput) + "," +
+                     "r=" + AUTHENTICATE.Extra.nonce
+
+AuthMessage = ClientFirstBare + "," + ServerFirst + "," +
+              ClientFinalNoProof
+```
+
+###### ClientProof
+
+`ClientProof` is computed by the client during the authentication exchange as follows:
+
+{align="left"}
+```javascript
+SaltedPassword  = KDF(Normalize(password), salt, params...)
+ClientKey       = HMAC(SaltedPassword, "Client Key")
+StoredKey       = SHA256(ClientKey)
+ClientSignature = HMAC(StoredKey, AuthMessage)
+ClientProof     = ClientKey XOR ClientSignature
+```
+
+The `ClientProof` is then sent to the server, base64-encoded, via the `AUTHENTICATE.Signature` argument.
+
+The server verifies the `ClientProof` by computing the `RecoveredStoredKey` and comparing it to the actual `StoredKey`:
+
+{align="left"}
+```javascript
+ClientSignature    = HMAC(StoredKey, AuthMessage)
+RecoveredClientKey = ClientSignature XOR ReceivedClientProof
+RecoveredStoredKey = SHA256(RecoveredClientKey)
+```
+
+Note that the client MAY cache the `ClientKey` and `StoredKey` (or just `SaltedPassword`) to avoid having to perform the expensive KDF computation for every authentication exchange. Storing these values securely on the client is outside the scope of this document.
+
+###### ServerSignature
+
+`ServerSignature` is computed by the server during the authentication exchange as follows:
+
+{align="left"}
+```javascript
+ServerSignature = HMAC(ServerKey, AuthMessage)
+```
+
+The `ServerSignature` is then sent to the client, base64-encoded, via the `WELCOME.Details.authextra.verifier` attribute.
+
+The client verifies the `ServerSignature` by computing it and comparing it with the `ServerSignature` sent by the server:
+
+{align="left"}
+```javascript
+ServerKey       = HMAC(SaltedPassword, "Server Key")
+ServerSignature = HMAC(ServerKey, AuthMessage)
+```
+
+##### Key Derivation Functions
+
+SCRAM uses a password-based key derivation function (KDF) to hash user passwords. WAMP-SCRAM supports both [Argon2](https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2/) and [PBKDF2](https://tools.ietf.org/html/rfc2898) as the KDF. Argon2 is recommended because of its memory hardness and resistance against GPU hardware. PBKDF2, which does not feature memory hardness, is also supported for applications that are required to use primitives currently approved by cryptographic standards.
+
+The following table maps the `CHALLENGE.Details.kdf` type string to the corresponding KDF.
+
+KDF type string | Function
+--------------- | --------
+`"argon2id13"` | Argon2id variant of Argon2, version 1.3
+`"pbkdf2"`      | PBKDF2
```

Comment:
Is there a version 1.9 or similar? The python library `argon2_cffi` refers to its latest argon2id hash as "v=19" is why I ask..

Reply:
As I've mentioned already in another related issue somewhere, 19 is the decimal representation of the hexadecimal number 0x13. Argon2 has chosen to use hexadecimal numbers to represent its version number.

Reply:
Ahh, sorry I missed that ;)

Reply:
As I am finally going over the PR, wondering, if `id==0x13` corresponds to `version==1.9`, then shouldn't the description `Argon2id variant of Argon2, version 1.3` read => `version 1.9`?

Reply:
From the [latest Argon2 IETF Draft](https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2):

> This document corresponds to version 1.3 of the Argon2 hash function.

>Version number v is one byte 0x13

If you print 0x13 as decimal, then you'll see 19.

The Argon2 authors decided to use hexadecimal representation for their versioning scheme. Perhaps they wanted it this way so that the second nibble changes only if the major version number changes.

Reply:
so actually, the version byte is 2 digit fixed point BCD. wtf? anyways, whatever;)

Reply:
Ah yes, BDC. That's the term I was looking for. The other alternatives would have been floating point and base-10 fixed point. With base-10 fixed point, you need to divide by 10 to obtain the major version number. With BCD, you only need to do a right shift. The latter might explain their motivation.